Introduce captionDisplayMode option to WebKitTestRunner

cola119 opened the pull request:
#### 2e6a6bb78167b54325e13d2f73d97608df4c65c7
<pre>
Introduce captionDisplayMode option to WebKitTestRunner
<a href="https://bugs.webkit.org/show_bug.cgi?id=261460">https://bugs.webkit.org/show_bug.cgi?id=261460</a>

Reviewed by Eric Carlson.

As reported in <a href="https://bugs.webkit.org/show_bug.cgi?id=179370#c2">https://bugs.webkit.org/show_bug.cgi?id=179370#c2</a>,
running WPT for WebVTT in run-webkit-tests results in a timeout.
This is due to the caption mode of WebKitTestRunner being defaulted to `ForceOnly`,
under which conditions the captions never appear and the tests stop.
This is not a problem for LayoutTests other than WPT because we can adjust
the caption mode with `window.internals.setCaptionDisplayMode(&apos;Automatic&apos;)`
to ensure correct caption display during testing.

To resolve this issue, this change introduces a new option `captionDisplayMode`
to WebKitTestRunner that makes the caption mode configurable instead of
using the internal methods in the window object.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/embedded-content/media-elements/track/track-element/track-text-track-cue-list-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/embedded-content/media-elements/track/track-element/track-text-track-cue-list.html:
* Source/WebCore/page/CaptionUserPreferences.h:
(WTF::EnumTraits&lt;WebCore::CaptionUserPreferences::CaptionDisplayMode&gt;::fromString):
* Source/WebKit/WebProcess/InjectedBundle/API/c/WKBundlePage.cpp:
(WKBundleSetCaptionDisplayMode):
* Source/WebKit/WebProcess/InjectedBundle/API/c/WKBundlePagePrivate.h:
* Source/WebKitLegacy/mac/WebView/WebFrame.mm:
(-[WebFrame _createCaptionPreferencesTestingModeToken]):
(-[WebFrame _setCaptionDisplayMode:]):
* Source/WebKitLegacy/mac/WebView/WebFrameInternal.h:
* Source/WebKitLegacy/mac/WebView/WebFramePrivate.h:
* Tools/DumpRenderTree/TestOptions.cpp:
(WTR::TestOptions::keyTypeMapping):
* Tools/DumpRenderTree/TestOptions.h:
(WTR::TestOptions::captionDisplayMode const):
* Tools/DumpRenderTree/mac/DumpRenderTree.mm:
(resetWebViewToConsistentState):
(runTest):
* Tools/WebKitTestRunner/InjectedBundle/InjectedBundle.cpp:
(WTR::InjectedBundle::beginTesting):
* Tools/WebKitTestRunner/TestInvocation.cpp:
(WTR::TestInvocation::createTestSettingsDictionary):
* Tools/WebKitTestRunner/TestOptions.cpp:
(WTR::TestOptions::defaults):
(WTR::TestOptions::keyTypeMapping):
* Tools/WebKitTestRunner/TestOptions.h:
(WTR::TestOptions::captionDisplayMode const):

Canonical link: <a href="https://commits.webkit.org/268073@main">https://commits.webkit.org/268073@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/57043c4e93599302fc75190a5f42d2cd43822453

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/18597 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/18935 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/19537 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/20457 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/17402 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/18793 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/22245 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/19077 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/19255 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/18822 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/18983 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/16189 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/21333 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/16205 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/16951 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/23399 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/17226 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/17121 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/21294 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/17718 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/15019 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/16772 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/16783 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4417 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/21140 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/17557 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

eric-carlson commented on Source/WebKitLegacy/mac/WebView/WebFrame.mm:
```diff
@@ -1911,6 +1913,18 @@ - (void)_setTextAutosizingWidth:(CGFloat)width
+{
+    auto* frame = core(self);
+    auto* page = frame->page();
+    if (!page)
```
Nit: `frame`  isn't necessary - `auto* page = core(self)->page()`

cola119 replied:
Done
https://github.com/WebKit/WebKit/compare/4a4cd4ccd21b3b3fd1d1c8495b9da5a0aa5b222c..9beb9395a825c5d381102c6fe0a5aeadf75866e7

eric-carlson reviewed (approved):
What a great change, thank you!

eric-carlson commented on Tools/DumpRenderTree/mac/DumpRenderTree.mm:
```diff
@@ -1827,6 +1827,13 @@ static void resetWebViewToConsistentState(const WTR::TestOptions& options, Reset
+    if (!options.captionDisplayMode().empty())
+        [mainFrame _setCaptionDisplayMode:[NSString stringWithUTF8String:options.captionDisplayMode().c_str()]];
+    else
+        [mainFrame _setCaptionDisplayMode:@"forcedOnly"];
```
Don't you want `@"forcedonly"`?

cola119 replied:
Done https://github.com/WebKit/WebKit/compare/322a92c6033af72d71b892d99ac0fe96aa933cbc..57043c4e93599302fc75190a5f42d2cd43822453

cola119 committed:
Introduce captionDisplayMode option to WebKitTestRunner
https://bugs.webkit.org/show_bug.cgi?id=261460

Reviewed by Eric Carlson.

As reported in https://bugs.webkit.org/show_bug.cgi?id=179370#c2,
running WPT for WebVTT in run-webkit-tests results in a timeout.
This is due to the caption mode of WebKitTestRunner being defaulted to `ForceOnly`,
under which conditions the captions never appear and the tests stop.
This is not a problem for LayoutTests other than WPT because we can adjust
the caption mode with `window.internals.setCaptionDisplayMode('Automatic')`
to ensure correct caption display during testing.

To resolve this issue, this change introduces a new option `captionDisplayMode`
to WebKitTestRunner that makes the caption mode configurable instead of
using the internal methods in the window object.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/embedded-content/media-elements/track/track-element/track-text-track-cue-list-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/embedded-content/media-elements/track/track-element/track-text-track-cue-list.html:
* Source/WebCore/page/CaptionUserPreferences.h:
(WTF::EnumTraits<WebCore::CaptionUserPreferences::CaptionDisplayMode>::fromString):
* Source/WebKit/WebProcess/InjectedBundle/API/c/WKBundlePage.cpp:
(WKBundleSetCaptionDisplayMode):
* Source/WebKit/WebProcess/InjectedBundle/API/c/WKBundlePagePrivate.h:
* Source/WebKitLegacy/mac/WebView/WebFrame.mm:
(-[WebFrame _createCaptionPreferencesTestingModeToken]):
(-[WebFrame _setCaptionDisplayMode:]):
* Source/WebKitLegacy/mac/WebView/WebFrameInternal.h:
* Source/WebKitLegacy/mac/WebView/WebFramePrivate.h:
* Tools/DumpRenderTree/TestOptions.cpp:
(WTR::TestOptions::keyTypeMapping):
* Tools/DumpRenderTree/TestOptions.h:
(WTR::TestOptions::captionDisplayMode const):
* Tools/DumpRenderTree/mac/DumpRenderTree.mm:
(resetWebViewToConsistentState):
(runTest):
* Tools/WebKitTestRunner/InjectedBundle/InjectedBundle.cpp:
(WTR::InjectedBundle::beginTesting):
* Tools/WebKitTestRunner/TestInvocation.cpp:
(WTR::TestInvocation::createTestSettingsDictionary):
* Tools/WebKitTestRunner/TestOptions.cpp:
(WTR::TestOptions::defaults):
(WTR::TestOptions::keyTypeMapping):
* Tools/WebKitTestRunner/TestOptions.h:
(WTR::TestOptions::captionDisplayMode const):

Canonical link: https://commits.webkit.org/268073@main
